proposals/interface-types/CanonicalABI.md

```diff
@@ -0,0 +1,1205 @@
+    #
+    # If the first handle value is 0 then the offset parameter is in bytes from
+    # the start of the memory of `src`. The `len` parameter is the length, in
+    # units of $ty, of the buffer. The pointer must be well aligned.
+    #
+    # If the first handle value is not zero then TABLES.get_buffer will validate
+    # that the handle is indeed a valid buffer for `src` to reference. The
+    # `offset` and `len` parameters are in units of `$ty` relative to within the
+    # original buffer. These values are validated to ensure they fit within the
+    # original buffer
+    push-buffer<$ty> | pull-buffer<$ty> => {
+      assert(direction == "import")
+      handle = values.assert_next(i32)
+      offset = values.assert_next(i32)
+      len = values.assert_next(i32)
+      if handle == 0:
+        assert_or_trap(offset + len * size(direction, $ty) <= src.memory.len)
+        yield(Buffer(src, offset, len, ty))
+      else:
+        buffer = TABLES.get_buffer(handle, src, ty).clone()
+        assert_or_trap(offset + len <= buffer.len)
+        buffer.offset += offset * size(direction, $ty)
+        buffer.len = len
+        yield(buffer)
+    }
+```
+
+There's a few more missing pieces to fill in, namely the in-memory
+representation of types due to the representation of `list` as contiguous
+elements in-memory. For this we can first define the `size` and `align` of all
+types:
+
+```
+def size(direction, type):
+  match type
+    f32 => 4
+    f64 => 8
+    s8 => 1
+    u8 => 1
+    s16 => 2
+    u16 => 2
+    s32 => 4
+    u32 => 4
+    s64 => 8
+    u64 => 8
+    char => 4
+    string |
+    list $ty => size(record([i32, i32]))
+    flags $fields => num_packed_i32s($fields) * 4
+    record $fields =>
+      s = 0
+      for field in $fields
+        s = align_to(s, align(field))
+        s += size(direction, field)
+      align_to(s, align(record $fields))
+    variant $cases =>
+      s = 0
+      discrim = variant_discriminant($cases)
+      for case in $cases
+        if case:
+          s = max(s, size(direction, record([discim, case])))
+        else:
+          s = max(s, size(direction, discrim))
+      s
+    handle => size(i32)
+    push-buffer | pull-buffer =>
+      if direction == "import":
+        size(record([i32, i32, i32]))
+      else:
+        size(i32)
+
+def align(type):
+  match type
+    f32 = 4
+    f64 = 8
+    s8 = 1
+    u8 = 1
+    s16 = 2
+    u16 = 2
+    s32 = 4
+    u32 = 4
+    s64 = 8
+    u64 = 8
+    char = 4
+    string = 4
+    list $ty = 4
+    flags $fields = 4
+    record $fields =
+      a = 1
+      for field in $fields
+        a = max(a, align(field))
+      a
+    variant $cases =
+      a = align(variant_discriminant($cases))
+      for case in $cases
+        if case:
+          a = max(a, align(case))
+      a
+    handle => align(i32)
+    push-buffer | pull-buffer => align(i32)
+
+def variant_discriminant(cases):
+  match cases.len() {
+    0 => unreachable,
```

Comment:
The explainer currently doesn’t forbid uninhabited enums, so either it or this assumption is wrong. A semi-reasonable type where we might hit this case is `Result<u8, !>`.